### PR TITLE
Multi operation

### PR DESCRIPTION
**Pull request by sgrove**

### CODE REVIEW

File: src/Explorer.js [hunk 1]
```diff
@@ -495,11 +502,16 @@ class ArgView extends React.PureComponent<ArgViewProps, ArgViewState> {
         setArgValue={this._setArgValue}
         getDefaultScalarArgValue={this.props.getDefaultScalarArgValue}
         makeDefaultArg={this.props.makeDefaultArg}
+        onRunOperation={event => this.props.onRunOperation()}
```
Comment (dwwoelfel) [inline src/Explorer.js]: Create `this._handleRunOperation` so it doesn't create a new function on each re-render. Can help with performance if it's using PureComponent.

File: src/Explorer.js [hunk 2]
```diff
@@ -495,11 +502,16 @@ class ArgView extends React.PureComponent<ArgViewProps, ArgViewState> {
...
       />
     );
   }
 }
 
+const isRunShortcut = event => {
```
Comment (dwwoelfel) [inline src/Explorer.js]: Function is more efficient and fewer characters

```
function isRunShortcut(event) {
  return event.metaKey && event.keyCode === 13;
}
```

Could you also add a comment about what keyCode 13 is.

File: src/Explorer.js [hunk 3]
```diff
@@ -599,6 +618,7 @@ class AbstractArgView extends React.PureComponent<AbstractArgViewProps, {}> {
               setArgValue={this.props.setArgValue}
               arg={arg}
               argValue={argValue}
+              onRunOperation={() => this.props.onRunOperation()}
```
Comment (dwwoelfel) [inline src/Explorer.js]: use class method

File: src/Explorer.js [hunk 4]
```diff
@@ -879,19 +903,63 @@ function defaultArgs(
...
+            kind: 'Field',
+            name: {kind: 'Name', value: fieldName},
+            arguments: [],
+          };
+        })
+      : [];
+
+    const subSelectionSet: SelectionSetNode = {
+      kind: 'SelectionSet',
+      selections: subFields || [],
```
Comment (dwwoelfel) [inline src/Explorer.js]: Do you need the `||`?

File: src/Explorer.js [hunk 5]
```diff
@@ -879,19 +903,63 @@ function defaultArgs(
...
-        kind: 'Field',
-        name: {kind: 'Name', value: this.props.field.name},
-        arguments: defaultArgs(
-          this.props.getDefaultScalarArgValue,
-          this.props.makeDefaultArg,
-          this.props.field,
-        ),
-      },
-    ]);
+  _addFieldToSelections = event => {
```
Comment (dwwoelfel) [inline src/Explorer.js]: The logic here is hard to follow because at a lot of subsequent steps, you have to check if you're trying to get all subselections or not.

Can you split the code into two cases, one where you press the alt key and one where you don't?

File: src/Explorer.js [hunk 6]
```diff
@@ -1096,113 +1195,142 @@ function memoizeParseQuery(query: string): DocumentNode {
...
-  _previousOperationDef: ?OperationDefinitionNode;
-  _getOperationDef = (parsedQuery: DocumentNode) => {
-    const operation = parsedQuery.definitions.find(
-      d =>
-        d.kind === 'OperationDefinition' &&
-        d.operation === this.props.operation,
-    );
-    const result = operation || {
+  _previousOperationDef: ?OperationDefinitionNode | ?FragmentDefinitionNode;
+  _temp = (): OperationDefinitionNode => {
```
Comment (dwwoelfel) [inline src/Explorer.js]: What does `_temp` do?

File: src/Explorer.js [hunk 7]
```diff
@@ -1096,113 +1195,142 @@ function memoizeParseQuery(query: string): DocumentNode {
...
+          )}
+        </div>
+
+        {Object.keys(fields || {})
           .sort()
           .map(fieldName => (
             <FieldView
               key={fieldName}
-              field={fields[fieldName]}
+              field={(fields || {})[fieldName]}
```
Comment (dwwoelfel) [inline src/Explorer.js]: Do you need `|| {}` here? Shouldn't be possible to get here with null fields.

File: src/Explorer.js [hunk 8]
```diff
@@ -1096,113 +1195,142 @@ function memoizeParseQuery(query: string): DocumentNode {
...
+                border: 'none',
+                borderBottom: '1px solid #888',
+                outline: 'none',
+                color: 'rgb(193, 42,80)',
+                width: `${Math.max(4, operationDisplayName.length)}ch`,
+              }}
+              autoComplete="false"
+              placeholder={`Name your ${operation}`}
+              value={this.props.name}
+              onKeyDown={event => {
```
Comment (dwwoelfel) [inline src/Explorer.js]: Make this and the onChange use methods on the class

File: src/Explorer.js [hunk 9]
```diff
@@ -1247,12 +1376,130 @@ class Explorer extends React.PureComponent<Props, State> {
...
+        : existingDefs.filter(def => {
+            if (def.kind === 'OperationDefinition') {
+              return def.operation === kind;
+            } else {
+              // Don't support adding fragments from explorer
+              return false;
+            }
+          });
+
+      const capitalize = string => {
```
Comment (dwwoelfel) [inline src/Explorer.js]: This can be moved out of the class

File: src/Explorer.js [hunk 10]
```diff
@@ -1269,42 +1516,111 @@ class Explorer extends React.PureComponent<Props, State> {
...
-            parsedQuery={parsedQuery}
-            onEdit={this._onEdit}
-            schema={schema}
-            getDefaultFieldNames={getDefaultFieldNames}
-            getDefaultScalarArgValue={getDefaultScalarArgValue}
-            makeDefaultArg={makeDefaultArg}
-          />
-        ) : null}
+        {relevantDefinitions.map((definition, index) => {
+          const operation:
```
Comment (dwwoelfel) [inline src/Explorer.js]: Why redefine definition to operation?

File: src/Explorer.js [hunk 11]
```diff
@@ -1269,42 +1516,111 @@ class Explorer extends React.PureComponent<Props, State> {
...
+          return (
+            <RootView
+              key={index}
+              fields={fields}
+              operation={operationKind}
+              name={operationName}
+              definition={definition}
+              onOperationRename={onOperationRename}
+              onTypeName={fragmentTypeName}
+              onEdit={newDefinition => {
```
Comment (dwwoelfel) [inline src/Explorer.js]: Make onEdit and onRunOperation use class methods.

File: src/Explorer.js (hunk 2 above)
Comment (sgrove) [reply src/Explorer.js]: I'm dubious this is a meaningful improvement, and hoisting is pretty annoying. Still, changed it to:

```
function isRunShortcut(event) {
  return event.metaKey && event.key === 'Enter';
}
```

File: src/Explorer.js (hunk 4 above)
Comment (sgrove) [reply src/Explorer.js]: Should be covered by the definition of `subFields`, was probably overly-defensive

File: src/Explorer.js (hunk 3 above)
Comment (sgrove) [reply src/Explorer.js]: Just replaced with `onRunOperation={this.props.onRunOperation}`

File: src/Explorer.js (hunk 5 above)
Comment (sgrove) [reply src/Explorer.js]: Done

File: src/Explorer.js (hunk 7 above)
Comment (sgrove) [reply src/Explorer.js]: Done

File: src/Explorer.js (hunk 10 above)
Comment (sgrove) [reply src/Explorer.js]: Fixed. I was (and still am, somewhat) confused about `definition` vs `operation`

File: src/Explorer.js (hunk 11 above)
Comment (sgrove) [reply src/Explorer.js]: `onEdit` relies on `parsedQuery` (which is inexpensively recoverable in a class method because of the memoization) and `operation` (which is in the local scope of the mapping function and not recoverable). How would you suggest hoisting `operation` out?

File: src/Explorer.js (hunk 7 above)
Comment (sgrove) [reply src/Explorer.js]: Flow will be unhappy since the type is `?GraphQLFieldMap<any, any>`, and it's pretty tough to get Flow to agree it's non-nullable (`queryFields/mutationFields/subscriptionFields/fragmentFields` could all be null)
